### PR TITLE
Add CodeStar Connection data source

### PR DESCRIPTION
**Pull request by shuheiktgw**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates https://github.com/hashicorp/terraform-provider-aws/issues/15453

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccDataSourceAwsCodeStarConnectionsConnection_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsCodeStarConnectionsConnection_ -timeout 180m
=== RUN   TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== PAUSE TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== RUN   TestAccDataSourceAwsCodeStarConnectionsConnection_tags
=== PAUSE TestAccDataSourceAwsCodeStarConnectionsConnection_tags
=== CONT  TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== CONT  TestAccDataSourceAwsCodeStarConnectionsConnection_tags
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_basic (37.21s)
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_tags (37.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	39.177s
```

Added CodeStar Connection data source. Thank you for your review!

### CODE REVIEW

File: aws/data_source_aws_codestarconnections_connection_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccDataSourceAwsCodeStarConnectionsConnection_basic(t *testing.T) {
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+	resourceName := "data.aws_codestarconnections_connection.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t); testAccPartitionHasServicePreCheck(codestarconnections.EndpointsID, t) },
+		Providers: testAccProviders,
```
Comment (bflad) [inline aws/data_source_aws_codestarconnections_connection_test.go:19]: We are currently in the process of adding `ErrorCheck` to each of the acceptance tests. Reference: https://github.com/hashicorp/terraform-provider-aws/issues/18175

Can you please add this here? Thank you!

```suggestion
		PreCheck:   func() { testAccPreCheck(t); testAccPartitionHasServicePreCheck(codestarconnections.EndpointsID, t) },
		ErrorCheck: testAccErrorCheck(t, codestarconnections.EndpointsID),
		Providers:  testAccProviders,
```

File: aws/data_source_aws_codestarconnections_connection_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAWSCodeStarConnectionsConnectionConfigBasic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccMatchResourceAttrRegionalARN(resourceName, "id", "codestar-connections", regexp.MustCompile("connection/.+")),
+					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "codestar-connections", regexp.MustCompile("connection/.+")),
+					resource.TestCheckResourceAttr(resourceName, "provider_type", codestarconnections.ProviderTypeBitbucket),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					resource.TestCheckResourceAttr(resourceName, "connection_status", codestarconnections.ConnectionStatusPending),
```
Comment (bflad) [inline aws/data_source_aws_codestarconnections_connection_test.go:28]: For data source testing, we prefer using `resource.TestCheckResourceAttrPair()` where possible to simplify it. 👍  Reference: https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/running-and-writing-acceptance-tests.md#data-source-acceptance-testing

```suggestion
					resource.TestCheckResourceAttrPair(dataSourceName, "arn", resourceName, "arn"),
					resource.TestCheckResourceAttrPair(dataSourceName, "connection_status", resourceName, "connection_status"),
					resource.TestCheckResourceAttrPair(dataSourceName, "id", resourceName, "id"),
					resource.TestCheckResourceAttrPair(dataSourceName, "provider_type", resourceName, "provider_type"),
					resource.TestCheckResourceAttrPair(dataSourceName, "name", resourceName, "name"),
					resource.TestCheckResourceAttrPair(dataSourceName, "tags.%", resourceName, "tags.%"),
```

File: aws/data_source_aws_codestarconnections_connection_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceAwsCodeStarConnectionsConnection_tags(t *testing.T) {
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+	resourceName := "data.aws_codestarconnections_connection.test"
```
Comment (bflad) [inline aws/data_source_aws_codestarconnections_connection_test.go:37]: Similarly:

```suggestion
	dataSourceName := "data.aws_codestarconnections_connection.test"
	resourceName := "aws_codestarconnections_connection.test"
```

File: aws/data_source_aws_codestarconnections_connection_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t); testAccPartitionHasServicePreCheck(codestarconnections.EndpointsID, t) },
+		Providers: testAccProviders,
```
Comment (bflad) [inline aws/data_source_aws_codestarconnections_connection_test.go:41]: Similarly:

```suggestion
		PreCheck:   func() { testAccPreCheck(t); testAccPartitionHasServicePreCheck(codestarconnections.EndpointsID, t) },
		ErrorCheck: testAccErrorCheck(t, codestarconnections.EndpointsID),
		Providers:  testAccProviders,
```

**Review by bflad — CHANGES_REQUESTED** [review]
Some minor testing updates, otherwise this is looking really good, thank you @shuheiktgw 🎉

**Comment by shuheiktgw** [comment]
Thanks, @bflad! Updated the tests, so would you review the PR again?

```
$ make testacc TESTARGS='-run=TestAccDataSourceAwsCodeStarConnectionsConnection_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsCodeStarConnectionsConnection_ -timeout 180m
go: downloading github.com/hashicorp/go-multierror v1.1.1
=== RUN   TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== PAUSE TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== RUN   TestAccDataSourceAwsCodeStarConnectionsConnection_tags
=== PAUSE TestAccDataSourceAwsCodeStarConnectionsConnection_tags
=== CONT  TestAccDataSourceAwsCodeStarConnectionsConnection_basic
=== CONT  TestAccDataSourceAwsCodeStarConnectionsConnection_tags
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_basic (45.73s)
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_tags (68.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	69.859s
```

**Review by bflad — APPROVED** [review]
Looks good to me 🚀  (Forgot to mention adding a changelog entry before 😅  will do so on merge)

Output from acceptance testing:

```
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_basic (19.66s)
--- PASS: TestAccDataSourceAwsCodeStarConnectionsConnection_tags (19.69s)

--- PASS: TestAccAWSCodeStarConnectionsConnection_disappears (12.08s)
--- PASS: TestAccAWSCodeStarConnectionsConnection_Basic (16.27s)
--- PASS: TestAccAWSCodeStarConnectionsConnection_Tags (35.69s)
```

**Comment by shuheiktgw** [comment]
Thank you for your review, @bflad!

**Pull request merged**